## Bridging Sequence and Structure to the Theory of Biology

My name is Joel Berendzen. I'm a computational biologist and biophysicist. I write **scalable, well-engineered software** that uses models from **physical chemistry, evolution, and ecology** to put **biological data in context**.

This is an exciting time to be a scientist, because a Theory of Biology that is emerging today in much the same way that the Theory of Physics emerged in the 1930's.  Biology has **evolutionary theory** (in its modern form of molecular phylogeny) at the quantum end of its scale.  At the cosmic scale lies **ecological theory** with roots in island ecology.  At both ends, the theory is quantitative and makes testable predictions with several significant figures of accuracy. In between, things get messy, and a large community of scientists works at building a bridge, making contact with physics, chemistry, and mathematics throughout the span.  The tools to build that bridge are increasingly embodied as code.

This page gives an overview of my code repositories, as well as some other places where I've contributed ideas and written code:

- [azulejo](https://github.com/joelb123/azulejo) combines guilt-by-profiling (genome synteny) and guilt-by-association (phylogeny) to create pangenomic collections of gene families. 
- [Sequedex](https://sequedex.lanl.gov) is award-winning software that uses scalable signature methods to classify short DNA sequences as to where they come from and what they do.  Sequedex is mostly used in metagenomics and surveillance for emergent infectious diseases. The Sequedex open-source repository is [here](https://github.com/lanl/sequedex-core).  
- [SOLVE](https://solve.lanl.gov) is award-winning software that helps automate the problem of phasing X-ray crystal structures of proteins.  It calculates a statistic that acts very much like autofocus on a camera.  SOLVE is closed-source.

